### PR TITLE
Creates Torch-friendly Event class and adds Stream tracking to autograd

### PR DESCRIPTION
**Pull request by mruberry**
Resubmission of #23424 because previous PR was borked.

### CODE REVIEW

**Comment by mruberry** [comment]
Sorry for the noise, Ed, last PR was seriously borked. Just need a stamp on this new version of it (code is the same).

**Comment by mruberry** [comment]
Rocm failure appears unrelated.

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h [hunk 1]
```diff
@@ -94,6 +94,97 @@ struct HIPGuardImplMasqueradingAsCUDA final : public c10::impl::DeviceGuardImplI
...
+    auto hip_event = static_cast<hipEvent_t>(event);
+    int orig_device;
+    try {
+      C10_HIP_CHECK(hipGetDevice(&orig_device));
+      C10_HIP_CHECK(hipSetDevice(device_index));
+      C10_HIP_CHECK(hipEventDestroy(hip_event));
+      C10_HIP_CHECK(hipSetDevice(orig_device));
+    } catch (...) {
+      // noexcept
+    }
```
[the remark this answers is not in the document]
Comment (mruberry) [reply aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h]: Great idea. Done.

File: c10/cuda/impl/CUDAGuardImpl.h [hunk 2]
```diff
@@ -58,6 +60,93 @@ struct CUDAGuardImpl final : public c10::impl::DeviceGuardImplInterface {
...
+  void destroyEvent(
+    void* event,
+    const DeviceIndex device_index) const noexcept override {
+    if (!event) return;
+    auto cuda_event = static_cast<cudaEvent_t>(event);
+    int orig_device;
+    cudaGetDevice(&orig_device);
+    cudaSetDevice(device_index);
+    cudaEventDestroy(cuda_event);
+    cudaSetDevice(orig_device);
```
[the remark this answers is not in the document]
Comment (mruberry) [reply c10/cuda/impl/CUDAGuardImpl.h]: Thanks for catching this. Done (with new macro).

File: aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h [hunk 3]
```diff
@@ -94,6 +94,97 @@ struct HIPGuardImplMasqueradingAsCUDA final : public c10::impl::DeviceGuardImplI
...
+      case EventFlag::PYTORCH_DEFAULT:
+      case EventFlag::HIP_EVENT_DISABLE_TIMING:
+        hip_flag = hipEventDisableTiming;
+        break;
+      case EventFlag::BACKEND_DEFAULT:
+      case EventFlag::HIP_EVENT_DEFAULT:
+        hip_flag = hipEventDefault;
+        break;
+      default:
+        AT_ERROR("HIP event received unknown flag");
```
[the remark this answers is not in the document]
Comment (mruberry) [reply aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h]: Updated to replace the AT_ checks in this and the CUDAGuardImpl file.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**